File: app/src/main/java/lt/tesonet/ui/servers/ServersActivity.java
package lt.tesonet.ui.servers;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.DividerItemDecoration;
import android.support.v7.widget.RecyclerView;
import android.view.View;
import android.widget.TextView;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;
import lt.tesonet.R;
import lt.tesonet.model.DatabaseUtil;
import lt.tesonet.model.Server;
import lt.tesonet.ui.login.LogInActivity;

public class ServersActivity extends AppCompatActivity {

    @BindView(R.id.server_list)
    RecyclerView list;
    @BindView(R.id.empty_text)
    TextView emptyText;

    private ServersAdapter adapter;

    private CompositeDisposable compositeDisposable = new CompositeDisposable();

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_servers);
        ButterKnife.bind(this);

        // init list
        DividerItemDecoration divider = new DividerItemDecoration(this, DividerItemDecoration.VERTICAL);
        adapter = new ServersAdapter();
        list.setAdapter(adapter);
        list.addItemDecoration(divider);

        loadData();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        compositeDisposable.clear();
    }

    @Override
    public void startActivity(Intent intent) {
        super.startActivity(intent);
        overridePendingTransition(R.anim.enter_new, R.anim.exit_old);
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(R.anim.enter_old, R.anim.exit_new);
    }

    private void showNoData() {
        list.setVisibility(View.GONE);
        emptyText.setVisibility(View.VISIBLE);
    }

    private void fillListWithData(List<Server> servers) {
        adapter.setData(servers);
        adapter.notifyDataSetChanged();
    }

    private void loadData() {
        compositeDisposable.add(DatabaseUtil.getServerDatabase(this).getDao()
                .getAllServers()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(servers -> {
                    if (servers.size() > 0) {
                        fillListWithData(servers);
                    } else {
                        showNoData();
                    }
                }, throwable -> showNoData()));
    }

    @OnClick(R.id.log_out)
    public void onLogOutClicked() {
        Intent i = new Intent(this, LogInActivity.class);
        startActivity(i);
        finish();
    }

}

File: app/src/main/java/com/example/android_party/ui/servers/ServersActivity.kt
package com.example.android_party.ui.servers

import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.android_party.R
import com.example.android_party.databinding.ServersActivityBinding
import com.example.android_party.injection.ViewModelFactory
import com.example.android_party.ui.login.LoginActivity

class ServersActivity : AppCompatActivity() {

    private lateinit var binding: ServersActivityBinding
    private lateinit var viewModel: ServersViewModel

    private var serversListAdapter: ServersListAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.servers_activity)
        viewModel =
            ViewModelProviders.of(this, ViewModelFactory()).get(ServersViewModel::class.java)
        viewModel.performLogout.observe(this, { value -> if (value) logout() })
        viewModel.loadingError.observe(this, { value ->
            showErrorToast(value)
            finish()
        })
        viewModel.loadingSuccess.observe(this, { value -> serversListAdapter?.updateList(value) })
        binding.viewModel = viewModel
        setUpServersRecyclerView(binding.serversRecyclerView)
    }

    private fun setUpServersRecyclerView(recyclerView: RecyclerView) {
        serversListAdapter = ServersListAdapter(this)
        recyclerView.layoutManager = LinearLayoutManager(this)
        recyclerView.adapter = serversListAdapter
    }

    private fun logout() {
        startActivity(Intent(this, LoginActivity::class.java))
        finish()
    }

    private fun showErrorToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }
}